Hook up IPC for Web Extension Storage API

kiaraarose opened the pull request:
#### 34adc76047884923a0638e2fcd01c41ea44daaff
<pre>
Hook up IPC for Web Extension Storage API
<a href="https://bugs.webkit.org/show_bug.cgi?id=267377">https://bugs.webkit.org/show_bug.cgi?id=267377</a>
<a href="https://rdar.apple.com/120811316">rdar://120811316</a>

Reviewed by Timothy Hatcher.

This patch implements the first step to supporting the Web Extension Storage API by hooking up IPC.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionConstants.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h.
Renamed since it now contains other constants not related to DNR.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::storage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm: Added.
(WebKit::WebExtensionAPIStorageArea::isPropertyAllowed):
(WebKit::WebExtensionAPIStorageArea::get):
(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
(WebKit::WebExtensionAPIStorageArea::set):
(WebKit::WebExtensionAPIStorageArea::remove):
(WebKit::WebExtensionAPIStorageArea::onChanged):
(WebKit::WebExtensionAPIStorageArea::clear):
(WebKit::WebExtensionAPIStorageArea::setAccessLevel):
(WebKit::WebExtensionAPIStorageArea::quotaBytes):
(WebKit::WebExtensionAPIStorageArea::quotaBytesPerItem):
(WebKit::WebExtensionAPIStorageArea::maxItems):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerHour):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerMinute):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm: Added.
(WebKit::WebExtensionAPIStorage::local):
(WebKit::WebExtensionAPIStorage::session):
(WebKit::WebExtensionAPIStorage::sync):
(WebKit::WebExtensionAPIStorage::onChanged):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDeclarativeNetRequest.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h: Copied from Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h: Added.
(WebKit::WebExtensionAPIStorageArea::WebExtensionAPIStorageArea):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorage.idl:
Renamed from Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl: Added.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/272992@main">https://commits.webkit.org/272992@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/001f8f391860d156da1214bb47f9cda685c37cba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33802 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12582 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35748 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36422 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30648 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34857 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9732 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29720 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34286 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/10647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30144 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9266 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30164 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37740 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/30718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30484 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35481 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9504 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7435 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33392 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11283 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10081 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4365 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10293 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
```diff
@@ -0,0 +1,47 @@
+    [RaisesException] void getBytesInUse([Optional, NSObject, DOMString] any items, [Optional, CallbackHandler] function callback);
+    [RaisesException] void set([NSDictionary] any items, [Optional, CallbackHandler] function callback);
+    [RaisesException] void remove([NSObject] any items, [Optional, CallbackHandler] function callback);
+    void clear([Optional, CallbackHandler] function callback);
```
forgot to add `setAccessLevel`. will do now

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -0,0 +1,77 @@
+
+    Util::loadAndRunExtension(storageManifest, @{ @"background.js": backgroundScript });
+}
+
```
remove extra newline

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionConstants.h:
```diff
@@ -0,0 +1,52 @@
+
+static constexpr NSUInteger webExtensionStorageAreaSyncMaxItems = 512;
+static constexpr NSUInteger webExtensionStorageAreaSyncMaxWriteOperationsPerHour = 1800;
+static constexpr NSUInteger webExtensionStorageAreaSyncMaxWriteOperationsPerMinute = 120;
```
These should use `size_t` to be consistent with the rest, and spell out "Maximum" like the others as well.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    if ([items isKindOfClass:NSString.class]) {
+        NSString *key = (NSString *)items;
```
`auto *key = (NSString *)items`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -0,0 +1,271 @@
+    }
+}
+
+void WebExtensionAPIStorageArea::getBytesInUse(NSObject *items, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
Should `items` be `keys` here?

xeenon replied:
Also `id` instead of `NSObject`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -0,0 +1,271 @@
+    if (items && !isString && !isArray && isNull) {
+        *outExceptionString = toErrorString(nil, @"items", @"it is expected to be either a string, array, or null");
+        return;
+    }
```
To keep the error consistent, use `validateObject`.

```
if (items && !validateObject(items, @"keys", [NSOrderedSet orderedSetWithObjects:NSString.class, @[ NSString.class ], NSNull.class, nil], outExceptionString))
    return;
```

kiaraarose replied:
forgot i could use NSOrderedSet for multiple input types! makes this much simpler. thanks!

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -0,0 +1,271 @@
+    callback->call();
+}
+
+void WebExtensionAPIStorageArea::remove(NSObject *items, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
`id` instead of `NSObject`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -0,0 +1,271 @@
+    if (!isArray && !isString) {
+        *outExceptionString = toErrorString(nil, @"items", @"it is expected to be either a string or an array");
+        return;
+    }
```
`validateObject`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h:
```diff
@@ -0,0 +1,62 @@
+    RefPtr<WebExtensionAPIStorageArea> m_sync;
+
+    RefPtr<WebExtensionAPIEvent> m_onChanged;
+
```
Remove empty line.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
```diff
@@ -0,0 +1,82 @@
+private:
+    explicit WebExtensionAPIStorageArea(ForMainWorld forMainWorld, WebExtensionAPIRuntimeBase& runtime, WebExtensionContextProxy& context, StorageType type)
+    : WebExtensionAPIObject(forMainWorld, runtime, context)
+    , m_type(type)
```
Indent one more level.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
```diff
@@ -0,0 +1,82 @@
+
+    StorageType m_type;
+    RefPtr<WebExtensionAPIEvent> m_onChanged;
+
```
Remove empty line.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -0,0 +1,258 @@
+        return;
+    }
+
+    if (items && ![items isKindOfClass:NSNull.class]) {
```
should also update this method to use validateObject with NSSet

kiaraarose committed:
Hook up IPC for Web Extension Storage API
https://bugs.webkit.org/show_bug.cgi?id=267377
rdar://120811316

Reviewed by Timothy Hatcher.

This patch implements the first step to supporting the Web Extension Storage API by hooking up IPC.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionConstants.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h.
Renamed since it now contains other constants not related to DNR.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::storage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm: Added.
(WebKit::WebExtensionAPIStorageArea::isPropertyAllowed):
(WebKit::WebExtensionAPIStorageArea::get):
(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
(WebKit::WebExtensionAPIStorageArea::set):
(WebKit::WebExtensionAPIStorageArea::remove):
(WebKit::WebExtensionAPIStorageArea::onChanged):
(WebKit::WebExtensionAPIStorageArea::clear):
(WebKit::WebExtensionAPIStorageArea::setAccessLevel):
(WebKit::WebExtensionAPIStorageArea::quotaBytes):
(WebKit::WebExtensionAPIStorageArea::quotaBytesPerItem):
(WebKit::WebExtensionAPIStorageArea::maxItems):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerHour):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerMinute):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm: Added.
(WebKit::WebExtensionAPIStorage::local):
(WebKit::WebExtensionAPIStorage::session):
(WebKit::WebExtensionAPIStorage::sync):
(WebKit::WebExtensionAPIStorage::onChanged):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDeclarativeNetRequest.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h: Copied from Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h: Added.
(WebKit::WebExtensionAPIStorageArea::WebExtensionAPIStorageArea):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorage.idl:
Renamed from Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl: Added.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/272992@main
